src/intrinsics.cpp

```diff
@@ -0,0 +1,558 @@
+        break;
+    }
+    case ISPCIntrinsics::concat: {
+        assert(Type::Equal(argTypes[0], argTypes[1]));
+        retType = lGetDoubleVectorType(TYs[0]);
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(retType, hasUnnamedType) + "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::cmpxchg: {
+        retType = TYs[1];
+        name += origName + "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::extract: {
+        llvm::VectorType *vt = llvm::dyn_cast<llvm::VectorType>(TYs[0]);
+        Assert(vt);
+        retType = vt->getElementType();
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[0], hasUnnamedType);
+        if (argTypes.size() == 3) {
+            name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        }
+        break;
+    }
+    case ISPCIntrinsics::fence: {
+        retType = llvm::Type::getVoidTy(*g->ctx);
+        name += origName;
+        break;
+    }
+    case ISPCIntrinsics::insert: {
+        llvm::VectorType *vt = llvm::dyn_cast<llvm::VectorType>(TYs[0]);
+        Assert(vt);
+        retType = vt;
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[0], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::packmask: {
+        retType = llvm::Type::getInt64Ty(*g->ctx);
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[0], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::select: {
+        retType = TYs[1];
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::stream_load: {
+        retType = TYs[1];
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::stream_store: {
+        retType = llvm::Type::getVoidTy(*g->ctx);
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    default: {
+        Error(SourcePos(), "Unknown ISPC intrinsic \"%s\".", ISPCIntrinsicsNames[(unsigned)ID]);
+        return nullptr;
+    }
+    }
+    llvm::FunctionType *funcType = llvm::FunctionType::get(retType, TYs, false);
+    auto p = llvm::cast<llvm::Function>(M->getOrInsertFunction(name, funcType).getCallee());
+    return p;
+}
+
+static std::vector<llvm::Type *> lDeductArgTypes(llvm::Intrinsic::ID ID, ExprList *args) {
+    std::vector<llvm::Type *> exprType;
+    // TODO: not all intrinsics are overloaded, so we need to handle them also?
+    if (llvm::Intrinsic::isOverloaded(ID)) {
+        Assert(args);
+        std::vector<int> nInits(args->exprs.size());
```

Comment:
Here more comments about nInits are needed. Some example of deducing types would be useful.

Reply:
I actually removed this change from this PR as it doesn't relate to ISPC intrinsics at all. I am going to move it to PR #3119